<a href="https://colab.research.google.com/github/kim-jason/PowerliftingDataset/blob/main/Powerlifting_Data_Science_Project.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Powerlifting Dataset EDA

Name: Jason Kim  <br>
Email: jason.s.kim20@gmail.com <br>
Linkedin: https://www.linkedin.com/in/jasonks/

## Introduction

### Overview of notebook


The purpose of this notebook is to analyze a dataset consisting of powerlifting meets in order to identify trends and insights by age, country, federation, and other provided and generated features. <br>

We will be using the CSV dataset from [OpenPowerlifting](https://openpowerlifting.gitlab.io/opl-csv/introduction.html), an open source site that regularly compiles official results from powerlifting competitions around the world. <br>

### Overview of powerlifting

Powerlifting is a sport consisting of three specific lifts: squat, bench press, and deadlift. The sum of the best lift in each discipline determines the winner in a competition. Lifters compete in given age groups and weight categories. 

 <!-- [![Everything Is AWESOME](http://i.imgur.com/Ot5DWAW.png)](https://youtu.be/StTqXEQ2l-Y?t=35s "Everything Is AWESOME") -->

 TO DO: Add description fo three

 


## Data import

In [1]:
# Import libraries used in the notebook
import pandas as pd
from datetime import datetime, timedelta
import math

In [2]:
# Download the raw CSV dataset (note I saved this CSV file to my personal Google Drive as it makes it faster to download and use in Google Colab)
! gdown --id 1b15_F2lH2sHI02m7F1HGQkDw3Bl25-X9

/usr/local/lib/python3.7/dist-packages/gdown/cli.py:131: FutureWarning: Option `--id` was deprecated in version 4.3.1 and will be removed in 5.0. You don't need to pass it anymore to use a file ID.
  category=FutureWarning,
Downloading...
From: https://drive.google.com/uc?id=1b15_F2lH2sHI02m7F1HGQkDw3Bl25-X9
To: /content/openpowerlifting-2022-08-06-98216f16.csv
100% 540M/540M [00:02<00:00, 257MB/s]


In [3]:
# Create dataframe with the raw CSV dataset
df = pd.read_csv('openpowerlifting-2022-08-06-98216f16.csv')
df.head()
print(len(df))

/usr/local/lib/python3.7/dist-packages/IPython/core/interactiveshell.py:2882: DtypeWarning: Columns (33,35,38) have mixed types.Specify dtype option on import or set low_memory=False.
  exec(code_obj, self.user_global_ns, self.user_ns)


2755243


In [4]:
# Describe the contents of the dataframe
df.describe().apply(lambda s: s.apply('{0:.1f}'.format))

Age BodyweightKg  Squat1Kg  Squat2Kg  Squat3Kg Squat4Kg  \
count  1720150.0    2719385.0  710550.0  702587.0  682372.0   7313.0   
mean        30.8         84.1     113.5      95.8      36.6     75.7   
std         13.2         22.7     140.2     164.7     193.0    182.8   
min          0.0         10.0    -555.0    -580.0    -600.5   -550.0   
25%         20.5         67.1      90.0      75.0    -155.0    -88.0   
50%         27.5         81.9     145.0     145.0     115.0    131.0   
75%         38.5         98.8     200.0     202.5     192.5    198.5   
max         98.0        300.0     555.0     577.5     595.0    592.4   

      Best3SquatKg   Bench1Kg   Bench2Kg   Bench3Kg  ... Deadlift1Kg  \
count    1821444.0  1158272.0  1144372.0  1107394.0  ...    810608.0   
mean         174.3       83.5       56.3      -16.5  ...       159.6   
std           69.0      101.3      126.2      141.0  ...       108.3   
min         -508.0     -515.0     -589.7     -575.0  ...      -461.0   
25%          122.5       55.0      -45.0     -135.0  ...       125.0   
50%          170.0      105.0       95.0      -60.0  ...       180.0   
75%          217.7      145.0      145.0      117.5  ...       224.5   
max          595.0      551.0      500.0      577.5  ...       450.0   

      Deadlift2Kg Deadlift3Kg Deadlift4Kg Best3DeadliftKg    TotalKg  \
count    795214.0    759616.0     19399.0       1970884.0  2564651.0   
mean        131.8        20.7        80.1           189.1      375.4   
std         157.4       211.4       187.3            62.2      206.0   
min        -502.5      -587.5      -500.0          -410.0        1.0   
25%         115.0      -202.5       -75.5           140.0      200.0   
50%         177.5       120.0       142.9           188.2      351.5   
75%         230.0       205.0       206.4           235.0      530.7   
max         467.5       487.5       440.5           487.5     1407.5   

            Dots      Wilks Glossbrenner   Goodlift  
count  2541034.0  2541034.0    2541034.0  2318067.0  
mean       273.0      272.0        256.5       63.9  
std        129.6      129.1        122.7       16.3  
min          0.7        0.7          0.6        0.5  
25%        141.6      141.0        133.5       52.2  
50%        294.6      293.6        274.1       63.6  
75%        371.1      369.5        350.0       75.2  
max        795.2      793.3        756.9      167.5  

[8 rows x 22 columns]

In [5]:
# Confirm datatypes of the features
df.dtypes

Name                 object
Sex                  object
Event                object
Equipment            object
Age                 float64
AgeClass             object
BirthYearClass       object
Division             object
BodyweightKg        float64
WeightClassKg        object
Squat1Kg            float64
Squat2Kg            float64
Squat3Kg            float64
Squat4Kg            float64
Best3SquatKg        float64
Bench1Kg            float64
Bench2Kg            float64
Bench3Kg            float64
Bench4Kg            float64
Best3BenchKg        float64
Deadlift1Kg         float64
Deadlift2Kg         float64
Deadlift3Kg         float64
Deadlift4Kg         float64
Best3DeadliftKg     float64
TotalKg             float64
Place                object
Dots                float64
Wilks               float64
Glossbrenner        float64
Goodlift            float64
Tested               object
Country              object
State                object
Federation           object
ParentFederation    

In [6]:
# Update Date from object to datetime
df['Date'] = pd.to_datetime(df['Date'])
df.dtypes

Name                        object
Sex                         object
Event                       object
Equipment                   object
Age                        float64
AgeClass                    object
BirthYearClass              object
Division                    object
BodyweightKg               float64
WeightClassKg               object
Squat1Kg                   float64
Squat2Kg                   float64
Squat3Kg                   float64
Squat4Kg                   float64
Best3SquatKg               float64
Bench1Kg                   float64
Bench2Kg                   float64
Bench3Kg                   float64
Bench4Kg                   float64
Best3BenchKg               float64
Deadlift1Kg                float64
Deadlift2Kg                float64
Deadlift3Kg                float64
Deadlift4Kg                float64
Best3DeadliftKg            float64
TotalKg                    float64
Place                       object
Dots                       float64
Wilks               

Here is a description of each feature in the provided dataset. Additional details for each feature can be found [here](https://openpowerlifting.gitlab.io/opl-csv/bulk-csv-docs.html). 

*   **Name**: Name of the lifter
*   **Sex**: Sex category in which the lifter competed (M, F, or Mx)
*   **Event**: The type of competition that the lifter entered (SBD, BD, SD, SB, S, B, D)
*   **Equipment**: Equipment category under which the lifts were performed (Raw, Wraps, Single-ply, Multi-ply, Unlimited, Straps)
*   **Age**: Age of the lifter on the start date of the meet
*   **AgeClass**: Age class in which the filter falls, based on exact age of the lifter on the day of the competition
*   **BirthYearClass**: Birth year class in which the filter falls
*   **Division**: Division of competition
*   **BodyweightKg**: Recorded bodyweight of the lifter at the time of competition
*   **WeightClassKg**: Weight class in which the lifter competed
*   **Squat1Kg, Bench1Kg, Deadlift1Kg**: First attempts for each of squat, bench, and deadlift, respectively
*   **Squat2Kg, Bench2Kg, Deadlift2Kg**: Second attempts for each of squat, bench, and deadlift, respectively
*   **Squat3Kg, Bench3Kg, Deadlift3Kg**: Third attempts for each of squat, bench, and deadlift, respectively
*   **Squat4Kg, Bench4Kg, Deadlift4Kg**: Fourth attempts for each of squat, bench, and deadlift, respectively
*   **Best3SquatKg, Best3BenchKg, Best3DeadliftKg**: Maximum of the first three successful attempts for the lift
*   **TotalKg**: Sum of Best3SquatKg, Best3BenchKg, and Best3DeadliftKg, if all three lifts were a success
*   **Place**: Recorded place of the lifter in the given division at the end of the meet
*   **Dots**: Dots points (formula used to assess strength of lifter)
*   **Wilks**: Wilks points (formula used to assess strength of lifter)
*   **Glossbrenner**: Glossbrenner points (formula used to assess strength of lifter)
*   **Goodlift**: IPF GL points (formula used to assess strength of lifter)
*   **Tested**: Yes if the lifter entered a drug-tested category, empty otherwise
*   **Country**: Home country of the lifter
*   **State**: Home state/province/oblast/division/etc of the lifter
*   **Federation**: The federation that hosted the meet
*   **ParentFederation**: Topmost federation that sanctioned the meet, usually the international body
*   **Date**: Start date of the meet
*   **MeetCountry**: Country in which the meet was held
*   **MeetState**: State, province, or region in which the meet was held
*   **MeetName**: Name of the meet

## Data cleaning

### Missing data


We will go through the different features to handle missing data.

In [7]:
# See the prevalence of missing data within our current features
total = df.isnull().sum().sort_values(ascending=False)
percent = (df.isnull().sum()/df.isnull().count()).sort_values(ascending=False)
missing_data = pd.concat([total, percent], axis=1, keys=['Total', 'Percent'])

display(missing_data)

Total   Percent
Squat4Kg          2747930  0.997346
Bench4Kg          2736202  0.993089
Deadlift4Kg       2735844  0.992959
State             2245365  0.814943
Squat3Kg          2072871  0.752337
Squat2Kg          2052656  0.745000
Squat1Kg          2044693  0.742110
Deadlift3Kg       1995627  0.724302
Deadlift2Kg       1960029  0.711382
Deadlift1Kg       1944635  0.705794
Bench3Kg          1647849  0.598078
Bench2Kg          1610871  0.584657
Bench1Kg          1596971  0.579612
Country           1227151  0.445388
Age               1035093  0.375681
ParentFederation   982040  0.356426
BirthYearClass     968626  0.351557
Best3SquatKg       933799  0.338917
Best3DeadliftKg    784359  0.284679
AgeClass           783840  0.284490
MeetState          741829  0.269243
Tested             718448  0.260757
Goodlift           437176  0.158671
Best3BenchKg       330252  0.119863
MeetTown           327442  0.118843
Dots               214209  0.077746
Wilks              214209  0.077746
Glossbrenner       214209  0.077746
TotalKg            190592  0.069174
BodyweightKg        35858  0.013014
WeightClassKg       34749  0.012612
Division             1476  0.000536
MeetCountry             0  0.000000
Date                    0  0.000000
Name                    0  0.000000
Federation              0  0.000000
Place                   0  0.000000
Sex                     0  0.000000
Equipment               0  0.000000
Event                   0  0.000000
MeetName                0  0.000000

It is expected that there will be missing data for several features, such as the bench, squat, and deadlift features (some competitions do not include all three lifts), as well as the state and country features. 

One feature that seems fairly high in missing data is age. We will try to clean and impute data for this feature. This feature is important for several analyses, including progression over time by lifters, maximum lifts by age, etc. 

We will first remove rows that have ages less than 14. Firstly, according to the [2022 technical rulebook](https://www.powerlifting.sport/rules/codes/info/technical-rules) of the International Powerlifting Federation, the largest powerlifting federation in the world, the minimum age is 14. Secondly, although other federations permit "youth" or "junior" categories for those under 14, given the purpose of this dataset is to provide insights for adults and because there are significant physical differences between youths and adults given puberty, I will filter out those under 14.

In [8]:
# Create a copy of the original dataframe 
cleaned_df = df.copy(deep=True)

# Remove rows in which age is less than 14
cleaned_df = cleaned_df[(cleaned_df['Age'] >= 14) | (cleaned_df['Age'].isnull())]

# Visualize the missing rows for age
null_age_df = cleaned_df[cleaned_df['Age'].isnull()]
display(null_age_df.head(5))

Name Sex Event Equipment  Age AgeClass  \
266       Rosimeire Santos Souza   F     B       Raw  NaN    45-49   
267   João Vitor Rodrigues Claro   M     B       Raw  NaN    13-15   
269      Eduardo Haggar da Silva   M     B       Raw  NaN    13-15   
270        Murieu Moreira Moreno   M     B       Raw  NaN    13-15   
271  Gabriel Schineider da Silva   M     B       Raw  NaN    20-23   

    BirthYearClass  Division  BodyweightKg WeightClassKg  ...  Tested  \
266            NaN   F 45-49          56.0            56  ...     NaN   
267            NaN  MT 13-15          56.0            56  ...     NaN   
269            NaN  MT 13-15          60.0            60  ...     NaN   
270            NaN  MT 13-15          75.0            75  ...     NaN   
271            NaN        MJ          67.5          67.5  ...     NaN   

     Country  State  Federation  ParentFederation       Date  MeetCountry  \
266      NaN    NaN     CONBRAP               GPC 2017-05-13       Brazil   
267      NaN    NaN     CONBRAP               GPC 2017-05-13       Brazil   
269      NaN    NaN     CONBRAP               GPC 2017-05-13       Brazil   
270      NaN    NaN     CONBRAP               GPC 2017-05-13       Brazil   
271      NaN    NaN     CONBRAP               GPC 2017-05-13       Brazil   

     MeetState  MeetTown                                   MeetName  
266        NaN   Anhembi  Campeonato Paulista de Supino e Lev.Terra  
267        NaN   Anhembi  Campeonato Paulista de Supino e Lev.Terra  
269        NaN   Anhembi  Campeonato Paulista de Supino e Lev.Terra  
270        NaN   Anhembi  Campeonato Paulista de Supino e Lev.Terra  
271        NaN   Anhembi  Campeonato Paulista de Supino e Lev.Terra  

[5 rows x 41 columns]

DOWN

We will also check if there are any rows in which the Age does not fit the appropriate AgeClass and/or BirthYearClass. For example, a lifter with Age 14 can compete in the BirthYearClass of 13-15 but cannot compete in the BirthYearClass of 20-23.

We can see that there are a total of 196,458 rows with invalid AgeClass and/or BirthYearClass entries for the Age. This is 7% of the total dataset. As the AgeClass and BirthYearClass categories are variable by each competition and since we can artificially create our own AgeClass and BirthYearClass categories for analysis, we will set those as null. 

In [9]:
#DOWN

# Remove rows in which the Age does not fit the appropriate AgeClass and/or BirthYearClass
def verify_age(x):
  if not math.isnan(x['Age']):
    if len(str(x['AgeClass'])) > 0 and '-' in str(x['AgeClass']):
      min, max = x['AgeClass'].split('-')
      if not (float(x['Age']) >= float(min) and float(x['Age'] <= float(max))):
        return False
      else:
        return True
    elif len(str(x['BirthYearClass'])) > 0 and '-' in str(x['BirthYearClass']):
      min, max = x['BirthYearClass'].split('-')
      if not (float(x['Age']) >= float(min) and float(x['Age'] <= float(max))):
        return False
      else:
        return True
    else:
      return True
  else:
    return True

cleaned_df['Age_verified'] = cleaned_df.apply(verify_age, axis=1)
print(len(cleaned_df[cleaned_df['Age_verified'] == False]))
print(len(cleaned_df))

cleaned_df.loc[cleaned_df.Age_verified == False, ['AgeClass']] = None
cleaned_df.loc[cleaned_df.Age_verified == False, ['BirthYearClass']] = None

cleaned_df = cleaned_df.drop(['Age_verified'], axis=1)

196458
2728795


We will now attempt to impute missing data entries for age in two methods:


1.   Use a lifter's age from a different competition entry to calculate their age for other competitions (more accurate)
2.   Use the average of their AgeClass range or their BirthYearClass range (less accurate)



In [10]:
# First method for data imputation for age

# Use a lifter's age from a different competition entry to calculate their age for other competitions 
lifters_missing_age = cleaned_df[cleaned_df['Age'].isnull()][['Name', 'Date']]
lifters_missing_age = lifters_missing_age.rename(columns={'Date':'Date_missing'})
lifters_missing_age = lifters_missing_age.drop_duplicates(subset=['Name', 'Date_missing'])

# Sort original dataset to find a row with an existing Age and Date to calculate birthday
df = df.sort_values(['Age', 'Date'])
lifters_missing_age = pd.merge(lifters_missing_age, cleaned_df[['Name', 'Age', 'Date']].drop_duplicates(subset=['Name']), on='Name', how='inner')
lifters_missing_age = lifters_missing_age.rename(columns={'Age':'Age_existing', 'Date':'Date_existing'})
lifters_missing_age = lifters_missing_age[(lifters_missing_age['Age_existing'].notnull()) & (lifters_missing_age['Date_existing'].notnull())]
lifters_missing_age['Age_missing'] = lifters_missing_age['Date_missing'] - (lifters_missing_age['Date_existing'] - lifters_missing_age['Age_existing'].apply(lambda x: timedelta(seconds=x*(60*60*24*365))))
lifters_missing_age['Age_missing'] = lifters_missing_age['Age_missing'].apply(lambda x: round((x.total_seconds() / 60 / 60 / 24 / 365.25)*2)/2) # We round to the nearest 0.5 increment as the original dataset did this

display(lifters_missing_age)

Name Date_missing  Age_existing Date_existing  \
58       Gelson de Oliveira   2017-05-13          39.0    2021-07-09   
59       Gelson de Oliveira   2018-04-14          39.0    2021-07-09   
60       Gelson de Oliveira   2018-11-25          39.0    2021-07-09   
61       Gelson de Oliveira   2017-08-27          39.0    2021-07-09   
62       Gelson de Oliveira   2017-09-09          39.0    2021-07-09   
...                     ...          ...           ...           ...   
982863       Anna Śliwińska   2004-08-09          18.5    2001-11-28   
982865  Viktoriya Olenytsya   2004-08-09          18.5    2003-02-26   
982871           Rémi Bhang   2004-08-09          22.5    2007-06-12   
982877    Dmitriy Dvornikov   2004-08-09          19.5    2002-11-21   
982884    Sergey Sosnovskiy   2004-08-09          24.0    2013-06-13   

        Age_missing  
58             35.0  
59             35.5  
60             36.5  
61             35.0  
62             35.0  
...             ...  
982863         21.0  
982865         20.0  
982871         19.5  
982877         21.0  
982884         15.0  

[75272 rows x 5 columns]

In [11]:
# Merge with the original dataframe to add in the estimated ages
cleaned_df = pd.merge(cleaned_df, lifters_missing_age[['Name', 'Date_missing', 'Age_missing']], how='left', left_on=['Name', 'Date'], right_on=['Name', 'Date_missing'])
cleaned_df.loc[cleaned_df.Age.isnull() & cleaned_df.Date_missing.notnull() & cleaned_df.Age_missing.notnull(), ['Age']] = cleaned_df['Age_missing']
cleaned_df = cleaned_df.drop(['Date_missing', 'Age_missing'], axis=1)

# Remove any estimated ages that are inaccurate based on the provided AgeClass and BirthYearClass
cleaned_df['Age_verified'] = cleaned_df.apply(verify_age, axis=1)
cleaned_df.loc[cleaned_df.Age_verified == False, ['Age']] = None
cleaned_df = cleaned_df.drop(['Age_verified'], axis=1)

display(cleaned_df.head(5))

Name Sex Event Equipment   Age AgeClass BirthYearClass  \
0             Alona Vladi   F   SBD       Raw  33.0    24-34          24-39   
1      Galina Solovyanova   F   SBD       Raw  43.0    40-44          40-49   
2          Daniil Voronin   M   SBD       Raw  15.5     None           None   
3          Aleksey Krasov   M   SBD       Raw  35.0    35-39          24-39   
4  Margarita Pleschenkova   M   SBD       Raw  26.5    24-34          24-39   

  Division  BodyweightKg WeightClassKg  ...  Tested  Country  State  \
0        O         58.30            60  ...     Yes   Russia    NaN   
1       M1         73.10            75  ...     Yes   Russia    NaN   
2        T         67.40            75  ...     Yes   Russia    NaN   
3        O         66.65            75  ...     Yes   Russia    NaN   
4        O         72.45            75  ...     Yes   Russia    NaN   

   Federation  ParentFederation       Date  MeetCountry  MeetState  MeetTown  \
0         GFP               NaN 2019-05-11       Russia        NaN   Bryansk   
1         GFP               NaN 2019-05-11       Russia        NaN   Bryansk   
2         GFP               NaN 2019-05-11       Russia        NaN   Bryansk   
3         GFP               NaN 2019-05-11       Russia        NaN   Bryansk   
4         GFP               NaN 2019-05-11       Russia        NaN   Bryansk   

          MeetName  
0  Open Tournament  
1  Open Tournament  
2  Open Tournament  
3  Open Tournament  
4  Open Tournament  

[5 rows x 41 columns]

In [12]:
# Second method for data imputation for age

# Use the average of the lifters's AgeClass range or the BirthYearClass range to estimate remaining missing ages
def get_average_age(x):
  if math.isnan(x['Age']):
    if len(str(x['AgeClass'])) > 0 and '-' in str(x['AgeClass']):
      min, max = x['AgeClass'].split('-')
      average = (float(min) + float(max)) / 2.0
      return round(average * 2) / 2
    elif len(str(x['BirthYearClass'])) > 0 and '-' in str(x['BirthYearClass']):
      min, max = x['BirthYearClass'].split('-')
      average = (float(min) + float(max)) / 2.0
      return round(average * 2) / 2
    else:
      return x['Age']
  else:
    return x['Age']
cleaned_df['Age'] = cleaned_df.apply(get_average_age, axis=1)

display(cleaned_df)

Name Sex Event   Equipment   Age AgeClass  \
0                   Alona Vladi   F   SBD         Raw  33.0    24-34   
1            Galina Solovyanova   F   SBD         Raw  43.0    40-44   
2                Daniil Voronin   M   SBD         Raw  15.5     None   
3                Aleksey Krasov   M   SBD         Raw  35.0    35-39   
4        Margarita Pleschenkova   M   SBD         Raw  26.5    24-34   
...                         ...  ..   ...         ...   ...      ...   
2728790            Inna Orobets   F   SBD  Single-ply  45.0    45-49   
2728791             Ágnes Szabó   F   SBD  Single-ply  27.5    24-34   
2728792   Brenda van der Meulen   F   SBD  Single-ply  37.5    35-39   
2728793            Kelly Phasey   F   SBD  Single-ply  39.5     None   
2728794  Hildeborg Juvet Hugdal   F   SBD  Single-ply  32.5    24-34   

        BirthYearClass Division  BodyweightKg WeightClassKg  ...  Tested  \
0                24-39        O         58.30            60  ...     Yes   
1                40-49       M1         73.10            75  ...     Yes   
2                 None        T         67.40            75  ...     Yes   
3                24-39        O         66.65            75  ...     Yes   
4                24-39        O         72.45            75  ...     Yes   
...                ...      ...           ...           ...  ...     ...   
2728790          40-49     Open        109.30           84+  ...     Yes   
2728791          24-39     Open        113.22           84+  ...     Yes   
2728792          24-39     Open        110.79           84+  ...     Yes   
2728793           None     Open        124.53           84+  ...     Yes   
2728794          24-39     Open        123.00           84+  ...     Yes   

             Country  State  Federation  ParentFederation       Date  \
0             Russia    NaN         GFP               NaN 2019-05-11   
1             Russia    NaN         GFP               NaN 2019-05-11   
2             Russia    NaN         GFP               NaN 2019-05-11   
3             Russia    NaN         GFP               NaN 2019-05-11   
4             Russia    NaN         GFP               NaN 2019-05-11   
...              ...    ...         ...               ...        ...   
2728790      Ukraine    NaN         IPF               IPF 2016-11-14   
2728791      Hungary    NaN         IPF               IPF 2016-11-14   
2728792  Netherlands    NaN         IPF               IPF 2016-11-14   
2728793      England    NaN         IPF               IPF 2016-11-14   
2728794       Norway    NaN         IPF               IPF 2016-11-14   

         MeetCountry  MeetState  MeetTown                  MeetName  
0             Russia        NaN   Bryansk           Open Tournament  
1             Russia        NaN   Bryansk           Open Tournament  
2             Russia        NaN   Bryansk           Open Tournament  
3             Russia        NaN   Bryansk           Open Tournament  
4             Russia        NaN   Bryansk           Open Tournament  
...              ...        ...       ...                       ...  
2728790          USA        NaN   Orlando  World Open Championships  
2728791          USA        NaN   Orlando  World Open Championships  
2728792          USA        NaN   Orlando  World Open Championships  
2728793          USA        NaN   Orlando  World Open Championships  
2728794          USA        NaN   Orlando  World Open Championships  

[2728795 rows x 41 columns]

We can see that there are some estimated ages that are less than 14, which again is the minimum age to compete. This suggests that either the entries with missing ages had incorrect dates or that the entries with existing ages and dates are incorrect. We will assume the former and as the 9,330 estimated rows are only 0.3% of the remaining dataset of 2,728,795 rows, we will simply remove these rows from the remaining dataset. Note that this cleaned dataset will be used for analyses involving age and dates, but we can use the original dataset for other analyses that don't require accurate ages and dates.

We can confirm all data entries now have recorded or estimated ages.

In [13]:
# Remove data entries with estimated or recorded ages below the minimum threshold of 14
print(len(cleaned_df[cleaned_df['Age'] < 14]))
print(len(cleaned_df))
cleaned_df = cleaned_df[cleaned_df['Age'] >= 14]

# Confirm all remaining data entries have recorded or imputed ages
print(len((cleaned_df[cleaned_df['Age'].isnull()])))

9330
2728795
0


We will also remove estimated ages above 100 as it can be safely assumed those are not realistic ages for competitive powerlifters. Online research also suggests that the age 100 is an appropriate threshold, as [Edith Traina](https://barbend.com/100-year-old-edith-murway-traina-powerlifter-guinness/) was recently entered into the 2022 Guinness Book of Records as the oldest competitive female powerlifter ever. We can see that her entries are included below.

Below we can see several of the rows in which the age estimations are to be removed. For example, Gerald Foster is estimated to be 539.5 years old, as this data was imputed by taking the average of his AgeClass of 80-999. 

The age feature should now be cleaned.

In [14]:
cleaned_df = cleaned_df.sort_values(by='Age', ascending=False)
display(cleaned_df.head(5))

display(cleaned_df[cleaned_df['Name'] == 'Edith Traina'].head(5))

cleaned_df = cleaned_df[cleaned_df['Age'] <= 100]

Name Sex Event  Equipment    Age AgeClass BirthYearClass  \
2026130  Edward E. Brown   M     B        Raw  539.5   80-999         70-999   
2007821     Jack Stinson   M   SBD      Wraps  539.5   80-999         70-999   
2248944    Charles Vause   M     B        Raw  539.5   80-999         70-999   
2247171    Edwin Free Jr   M     B  Multi-ply  539.5   80-999         70-999   
105835   George Matthews   M     B        Raw  539.5   80-999         70-999   

                      Division  BodyweightKg WeightClassKg  ...  Tested  \
2026130  Amateur Masters 80-84         99.34           100  ...     Yes   
2007821  Amateur Masters 85-89         88.36            90  ...     Yes   
2248944          Masters 80-84         75.00            75  ...     Yes   
2247171          Masters 80-84         89.72            90  ...     Yes   
105835                 M 80-84         89.00            90  ...     NaN   

         Country  State  Federation  ParentFederation       Date  MeetCountry  \
2026130      USA     NY         RPS               NaN 2020-02-08          USA   
2007821      USA     NH         RPS               NaN 2015-12-05          USA   
2248944      USA     HI       WABDL               NaN 2018-07-01          USA   
2247171      USA     TN       WABDL               NaN 2005-11-16          USA   
105835       USA    NaN         APC              WUAP 2017-06-24          USA   

         MeetState     MeetTown                     MeetName  
2026130         VT   Fair Haven       Vermont Winter Classic  
2007821         VT   Fair Haven                   Power Wars  
2248944         HI          NaN  Sonny's 14th Annual Classic  
2247171         NV    Las Vegas          World Championships  
105835          GA  Gainesville        15th Annual Nationals  

[5 rows x 41 columns]

Name Sex Event Equipment   Age AgeClass BirthYearClass  \
2214450  Edith Traina   F     B       Raw  98.0   80-999         70-999   
2214011  Edith Traina   F    BD       Raw  97.0   80-999         70-999   
2215788  Edith Traina   F     D       Raw  96.0   80-999         70-999   
2213855  Edith Traina   F     B       Raw  96.0   80-999         70-999   
2214521  Edith Traina   F    BD       Raw  94.5   80-999         70-999   

        Division  BodyweightKg WeightClassKg  ...  Tested  Country  State  \
2214450     Open         67.50          67.5  ...     NaN      NaN    NaN   
2214011      SNR         75.00            75  ...     NaN      NaN    NaN   
2215788      SNR         75.00            75  ...     NaN      NaN    NaN   
2213855    95-99         67.13          67.5  ...     NaN      NaN    NaN   
2214521    90-94         75.00            75  ...     NaN      NaN    NaN   

         Federation  ParentFederation       Date  MeetCountry  MeetState  \
2214450        RAWU               NaN 2019-11-10          USA         FL   
2214011        RAWU               NaN 2018-09-15          USA         FL   
2215788        RAWU               NaN 2018-06-08          USA         FL   
2213855        RAWU               NaN 2017-12-30          USA         FL   
2214521        RAWU               NaN 2016-09-23          USA         FL   

           MeetTown                             MeetName  
2214450       Tampa            Tony Conyers Extravaganza  
2214011       Tampa                            Push/Pull  
2215788       Tampa  Grip it and Rip it Deadlift Contest  
2213855  Mount Dora                          Beasty Bash  
2214521       Tampa            Tony Conyers Extravaganza  

[5 rows x 41 columns]

Next, we will see if the BodyweightKg feature has any data that could be potentially inaccurate.

We can see that the minimum and maximum range for BodyweightKg of 10 kg and 300 kg (95 lbs and 309 lbs) can be assumed to be accurate as they are within reasonable boundaries of human bodyweight.

In [16]:
print(cleaned_df['BodyweightKg'].max())
print(cleaned_df['BodyweightKg'].min())

cleaned_df = cleaned_df.sort_values(by='BodyweightKg')
display(cleaned_df.head(20))
display(cleaned_df[cleaned_df['BodyweightKg'] == 300.0])
display(cleaned_df[cleaned_df['BodyweightKg'] == 10.0])

300.0
10.0


Name Sex Event   Equipment   Age AgeClass  \
1330257          R. Honahan   M   SBD  Single-ply  62.0    60-64   
2119154     Ziphokonke Cele   F   SBD       Wraps  18.0    18-19   
2673060      Artem Sultanov   M     B         Raw  14.0    13-15   
1218593     Angelia Aguirre   F     B   Multi-ply  14.0    13-15   
2024187       Dávid Barabás   M   SBD         Raw  29.0    24-34   
1577453    Gonçalo Teixeira   M     D         Raw  14.0    13-15   
951672            Erin Shaw   F   SBD       Wraps  14.0    13-15   
1076456      Brett Johns #2   M   SBD       Wraps  16.5    16-17   
2073569       Darnell Floyd   M     B         Raw  14.0    13-15   
2016718     Zachary Bradley   M    BD         Raw  14.0    13-15   
2673061  Nikita Korobenikov   M     B         Raw  14.0    13-15   
913378     Arseniy Lemeshko   M   SBD         Raw  16.0    16-17   
913393     Arseniy Lemeshko   M     S         Raw  16.0    16-17   
1660533    Andrey Torlevich   M     B         Raw  20.5    20-23   
1726806   Nikon Vternovskiy   M     B         Raw  14.5    13-15   
925353       Lev Mitrofanov   M     D         Raw  15.0    13-15   
445219      Catherine Gruss   F    BD         Raw  27.5    24-34   
925380       Lev Mitrofanov   M     B         Raw  15.0    13-15   
131589      Dinara Shpeyker   F   SBD         Raw  14.0    13-15   
771271     Brittany Trevino   F   SBD  Single-ply  18.5    18-19   

        BirthYearClass            Division  BodyweightKg WeightClassKg  ...  \
1330257          60-69               60-64         10.00            10  ...   
2119154            NaN                F-T3         18.00            44  ...   
2673060          14-18  Amateur Teen 14-15         22.50            52  ...   
1218593            NaN        F_TEM_1_AAPF         24.95            44  ...   
2024187          24-39        Amateur Open         24.95            52  ...   
1577453            NaN             T 13-15         25.00            52  ...   
951672           14-18                FT1R         25.00            44  ...   
1076456            NaN               Youth         25.30            48  ...   
2073569          14-18            RT 14-15         25.63            44  ...   
2016718          14-18  Amateur Teen 14-15         26.49            52  ...   
2673061          14-18  Amateur Teen 14-15         27.30            52  ...   
913378           14-18          Teen 16-17         28.00            52  ...   
913393           14-18          Teen 16-17         28.00            52  ...   
1660533          19-23             Juniors         28.50            53  ...   
1726806          14-18         Sub-Juniors         29.35            32  ...   
925353           14-18          Teen 13-15         29.40            52  ...   
445219           24-39                FR-O         29.40            44  ...   
925380           14-18          Teen 13-15         29.40            52  ...   
131589             NaN      Pro Teen 13-15         29.65            44  ...   
771271             NaN               Girls         29.66          44.2  ...   

         Tested       Country  State  Federation  ParentFederation       Date  \
1330257     NaN           NaN    NaN        USPF               IPF 1985-03-10   
2119154     NaN  South Africa    NaN         GPC               GPC 2019-11-10   
2673060     Yes        Russia    NaN         NAP               IPA 2015-10-16   
1218593     NaN           NaN    NaN         APF               WPC 2002-04-06   
2024187     Yes       Hungary    NaN         RPS               NaN 2017-08-12   
1577453     NaN      Portugal    NaN         WPC               WPC 2013-06-02   
951672      NaN           NaN    NaN         UPA               NaN 2013-04-20   
1076456     Yes     Australia     SA       AAPLF               IPF 1977-01-01   
2073569     Yes           USA     NC         AAU               NaN 2002-05-10   
2016718     Yes           NaN    NaN         RPS               NaN 2017-08-20   
2673061     Yes        Russia    NaN       

Name Sex Event Equipment   Age AgeClass BirthYearClass  Division  \
2166487  Les Wroe   M     D       Raw  33.0    24-34          24-39  Pro Open   

         BodyweightKg WeightClassKg  ...  Tested    Country  State  \
2166487         300.0           NaN  ...     NaN  Australia    NaN   

         Federation  ParentFederation       Date  MeetCountry  MeetState  \
2166487      MM-AUS                MM 2022-06-18    Australia        QLD   

          MeetTown    MeetName  
2166487  Fordsdale  Redemption  

[1 rows x 41 columns]

Name Sex Event   Equipment   Age AgeClass BirthYearClass  \
1330257  R. Honahan   M   SBD  Single-ply  62.0    60-64          60-69   

        Division  BodyweightKg WeightClassKg  ...  Tested  Country  State  \
1330257    60-64          10.0            10  ...     NaN      NaN    NaN   

         Federation  ParentFederation       Date  MeetCountry  MeetState  \
1330257        USPF               IPF 1985-03-10          USA         CA   

            MeetTown                      MeetName  
1330257  Los Angeles  Senior Olympics Powerlifting  

[1 rows x 41 columns]

Next, we will check that all competitors are competing in appropriate weight classes. For example, an athlete with a WeightClassKg of 75 should have a BodyweightKg less than or equal to 75; an athlete with a WeightClassKg of 100+ should have a BodyweightKg greater than or equal to 100.

In [17]:
def validate_weightclasskg(x):
  if len(str(x['WeightClassKg'])) > 0:
    if '+' in str(x['WeightClassKg']):
      if len(str(x['WeightClassKg'])) > 1:
        min, _ = x['WeightClassKg'].split('+')
        return float(x['BodyweightKg']) >= float(min)
      else:
        return True
    else:
      return float(x['BodyweightKg']) <= float(x['WeightClassKg'])
  else:
    return True

cleaned_df['Valid_WeightClassKg'] = cleaned_df.apply(validate_weightclasskg, axis=1)
display(cleaned_df[cleaned_df['Valid_WeightClassKg'] == False])

Name Sex Event   Equipment   Age AgeClass  \
1597334          Olga Igisheva   F   SBD  Single-ply  14.5    13-15   
919801   Anastasiya Novoselova   F     D         Raw  22.5    20-23   
919777   Anastasiya Novoselova   F     B         Raw  22.5    20-23   
832739        Maxim Novychenko   M     B         Raw  24.5    24-34   
1645051      Marina Evtushenko   F     B  Single-ply  41.5    40-44   
...                        ...  ..   ...         ...   ...      ...   
1124828          Emily Walther   F   SBD  Single-ply  14.0    13-15   
1484534        Samantha Füßler   F     D  Single-ply  14.0    13-15   
38109          Isabella Dupont   F     B         Raw  14.0    13-15   
1144634             P. McCarty   M   SBD  Single-ply  14.0    13-15   
1146134              T. Webber   M     B  Single-ply  14.0    13-15   

        BirthYearClass    Division  BodyweightKg WeightClassKg  ...  Country  \
1597334          14-18     Juniors          37.5           NaN  ...      NaN   
919801           19-23        Open          38.0           NaN  ...      NaN   
919777           19-23        Open          38.0           NaN  ...      NaN   
832739           24-39        Open          38.7           NaN  ...  Ukraine   
1645051          40-49   Masters 1          38.9           NaN  ...   Russia   
...                ...         ...           ...           ...  ...      ...   
1124828          14-18  Teen 14-15           NaN           90+  ...      USA   
1484534          14-18     Youth B           NaN           84+  ...  Germany   
38109            14-18  Teen 14-15           NaN           90+  ...      USA   
1144634          14-18  Teen 14-15           NaN           NaN  ...      NaN   
1146134            NaN      Novice           NaN          125+  ...      USA   

         State  Federation  ParentFederation       Date  MeetCountry  \
1597334    NaN         FPR               IPF 2004-02-13       Russia   
919801     NaN         RPU               NaN 2015-07-18       Russia   
919777     NaN         RPU               NaN 2015-07-18       Russia   
832739     NaN        WRPF              WRPF 2018-12-22      Belarus   
1645051    NaN         FPR               IPF 2014-02-14       Russia   
...        ...         ...               ...        ...          ...   
1124828    NaN         WPF               NaN 1999-12-17          USA   
1484534    NaN        BVDK               IPF 2016-12-03      Germany   
38109       RI         RAW               NaN 2014-03-15          USA   
1144634    NaN        USPF               IPF 1983-03-05          USA   
1146134    NaN        USPF               IPF 1996-05-02          USA   

         MeetState     MeetTown  \
1597334         BA  Mesyagutovo   
919801         KEM     Kemerovo   
919777         KEM     Kemerovo   
832739         NaN        Minsk   
1645051        YAN     Noyabrsk   
...            ...          ...   
1124828         NV    Las Vegas   
1484534        NaN   Greifswald   
38109           MA      Peabody   
1144634         AR    El Dorado   
1146134         OR     Portland   

                                                  MeetName  \
1597334           Bashkortostan Powerlifting Championships   
919801                                 Kemerovo Tournament   
919777                                 Kemerovo Tournament   
832739                               Adrenaline Grand Prix   
1645051  Yamalo-Nenets Autonomous Okrug Bench Press Сha...   
...                                                    ...   
1124828                                World Championships   
1484534               Deutsche Meisterschaft im Kreuzheben   
38109                                        March Madness   
1144634                          Greater Mr. Arkansas Meet   
1146134                               National Bench Press   

         Valid_WeightClassKg  
1597334                False  
919801                 False  
919777                 False  
832739                 False  
1645051                Fal